File: modules/desktop-ui/src/main/java/com/projecthub/ui/modules/dashboard/DashboardViewModel.java
package com.projecthub.ui.modules.dashboard;

import com.projecthub.base.dashboard.application.service.DashboardService;
import javafx.beans.property.IntegerProperty;
import javafx.beans.property.SimpleIntegerProperty;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.scene.chart.PieChart;
import org.springframework.stereotype.Component;

import java.util.stream.Collectors;

@Component
public class DashboardViewModel {

    private final DashboardService dashboardService;

    private final IntegerProperty totalUsers = new SimpleIntegerProperty();
    private final IntegerProperty totalProjects = new SimpleIntegerProperty();
    private final IntegerProperty totalTeams = new SimpleIntegerProperty();
    private final ObservableList<PieChart.Data> projectStatusDistribution = FXCollections.observableArrayList();
    private final ObservableList<RecentActivity> recentActivities = FXCollections.observableArrayList();


    public DashboardViewModel(DashboardService dashboardService) {
        this.dashboardService = dashboardService;
    }

    public IntegerProperty totalUsersProperty() {
        return totalUsers;
    }

    public IntegerProperty totalProjectsProperty() {
        return totalProjects;
    }

    public IntegerProperty totalTeamsProperty() {
        return totalTeams;
    }

    public ObservableList<PieChart.Data> getProjectStatusDistribution() {
        return projectStatusDistribution;
    }

    public ObservableList<RecentActivity> getRecentActivities() {
        return recentActivities;
    }

    public void loadDashboardData() {
        totalUsers.set(dashboardService.getTotalUsers());
        totalProjects.set(dashboardService.getTotalProjects());
        totalTeams.set(dashboardService.getTotalTeams());

        projectStatusDistribution.setAll(
            dashboardService.getProjectStatusDistribution().stream()
                .map(dto -> new PieChart.Data(dto.getStatus(), dto.getCount()))
                .collect(Collectors.toList())
        );

        recentActivities.setAll(
            dashboardService.getRecentActivities().stream()
                .map(dto -> new RecentActivity(dto.getTimestamp(), dto.getActivity(), dto.getUser()))
                .collect(Collectors.toList())
        );
    }

    public record RecentActivity(String timestamp, String activity, String user) {
    }
}

File: src/main/kotlin/com/projecthub/interfaces/compose/features/dashboard/viewmodel/DashboardViewModel.kt
package com.projecthub.ui.dashboard.viewmodel

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.projecthub.base.dashboard.application.service.DashboardService
import com.projecthub.base.project.mgmt.domain.service.ProjectService
import com.projecthub.base.task.infrastructure.service.TaskService
import com.projecthub.ui.dashboard.model.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch

class DashboardViewModel(
    private val dashboardService: DashboardService,
    private val projectService: ProjectService,
    private val taskService: TaskService,
    private val coroutineScope: CoroutineScope
) {
    var state by mutableStateOf(DashboardState())
        private set

    private val _events = MutableSharedFlow<DashboardEvent>()
    val events: SharedFlow<DashboardEvent> = _events.asSharedFlow()

    init {
        coroutineScope.launch {
            _events.collect { event ->
                handleEvent(event)
            }
        }
        loadDashboard()
    }

    private suspend fun handleEvent(event: DashboardEvent) {
        when (event) {
            is DashboardEvent.LoadDashboard -> loadDashboard()
            is DashboardEvent.LoadRecentActivities -> loadRecentActivities()
            is DashboardEvent.LoadProjectMetrics -> loadProjectMetrics(event.projectId)
            is DashboardEvent.ClearError -> clearError()
        }
    }

    private fun loadDashboard() {
        state = state.copy(isLoading = true)
        try {
            val metrics = DashboardMetrics(
                totalUsers = dashboardService.getTotalUsers(),
                totalProjects = dashboardService.getTotalProjects(),
                totalTeams = dashboardService.getTotalTeams(),
                projectStatusDistribution = dashboardService.getProjectStatusDistribution()
                    .map { ProjectStatusData(it.status, it.count) }
            )

            state = state.copy(
                metrics = metrics,
                isLoading = false
            )
            
            // Load recent activities after metrics
            loadRecentActivities()
        } catch (e: Exception) {
            state = state.copy(
                error = e.message ?: "Failed to load dashboard",
                isLoading = false
            )
        }
    }

    private fun loadRecentActivities() {
        try {
            val activities = dashboardService.getRecentActivities()
                .map { dto ->
                    RecentActivity(
                        timestamp = dto.timestamp,
                        activity = dto.activity,
                        user = dto.user,
                        type = when {
                            dto.activity.contains("created project") -> ActivityType.PROJECT_CREATED
                            dto.activity.contains("completed task") -> ActivityType.TASK_COMPLETED
                            dto.activity.contains("updated team") -> ActivityType.TEAM_UPDATED
                            dto.activity.contains("graded submission") -> ActivityType.SUBMISSION_GRADED
                            dto.activity.contains("joined") -> ActivityType.USER_JOINED
                            else -> ActivityType.PROJECT_CREATED
                        }
                    )
                }

            state = state.copy(recentActivities = activities)
        } catch (e: Exception) {
            state = state.copy(
                error = e.message ?: "Failed to load recent activities"
            )
        }
    }

    private fun loadProjectMetrics(projectId: String) {
        try {
            val tasks = taskService.getTasksByProjectId(projectId)
            
            val metrics = ProjectDashboardMetrics(
                totalTasks = tasks.size,
                completedTasks = tasks.count { it.status == "COMPLETED" },
                inProgressTasks = tasks.count { it.status == "IN_PROGRESS" },
                pendingTasks = tasks.count { it.status == "PENDING" }
            )

            state = state.copy(projectMetrics = metrics)
        } catch (e: Exception) {
            state = state.copy(
                error = e.message ?: "Failed to load project metrics"
            )
        }
    }

    private fun clearError() {
        state = state.copy(error = null)
    }

    fun emitEvent(event: DashboardEvent) {
        coroutineScope.launch {
            _events.emit(event)
        }
    }
}